src/platform_impl/linux/x11/util/cursor.rs

```diff
@@ -1,12 +1,63 @@
-use std::ffi::CString;
-use std::iter;
+use std::{ffi::CString, iter, slice};
 
 use x11rb::connection::Connection;
 
-use crate::window::CursorIcon;
+use crate::{cursor::CursorImage, window::CursorIcon};
 
 use super::*;
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct CustomCursorInternal(ffi::Cursor);
+
+impl CustomCursorInternal {
+    pub unsafe fn new(
+        xcursor: &ffi::Xcursor,
+        display: *mut ffi::Display,
+        image: &CursorImage,
+    ) -> Self {
+        unsafe {
+            let ximage = (xcursor.XcursorImageCreate)(image.width as i32, image.height as i32);
+            if ximage.is_null() {
+                panic!("failed to allocate cursor image");
+            }
+            (*ximage).xhot = image.hotspot_x;
+            (*ximage).yhot = image.hotspot_y;
+            (*ximage).delay = 0;
+
+            let dst =
+                slice::from_raw_parts_mut((*ximage).pixels, (image.width * image.height) as usize);
+            for (i, chunk) in image.rgba.chunks_exact(4).enumerate() {
+                dst[i] = (chunk[0] as u32) << 16
+                    | (chunk[1] as u32) << 8
+                    | (chunk[2] as u32)
+                    | (chunk[3] as u32) << 24;
+            }
+
+            let cursor = (xcursor.XcursorImageLoadCursor)(display, ximage);
+            (xcursor.XcursorImageDestroy)(ximage);
+            Self(cursor)
+        }
+    }
+
+    pub unsafe fn destroy(&self, xlib: &ffi::Xlib, display: *mut ffi::Display) {
+        unsafe {
+            (xlib.XFreeCursor)(display, self.0);
+        }
+    }
+}
```

Comment:
Rather than having an unsafe drop function like this, I would prefer for the `CustomCursorInternal` type to clone the `Arc<XConnection>` and use that in a safer `drop` implementation.

Reply:
Done.